flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/hybrid/HybridSourceReader.java

```diff
@@ -0,0 +1,213 @@
+    private long lastCheckpointId = -1;
+    private SourceReader<T, ? extends SourceSplit> currentReader;
+    // track last availability to resume reader after source switch
+    private CompletableFuture<Void> availabilityFuture;
+
+    public HybridSourceReader(
+            SourceReaderContext readerContext,
+            List<SourceReader<T, ? extends SourceSplit>> readers) {
+        this.readerContext = readerContext;
+        this.realReaders = readers;
+    }
+
+    @Override
+    public void start() {
+        setCurrentReader(0);
+    }
+
+    @Override
+    public InputStatus pollNext(ReaderOutput output) throws Exception {
+        InputStatus status = currentReader.pollNext(output);
+        if (status == InputStatus.END_OF_INPUT) {
+            // trap END_OF_INPUT if this wasn't the final reader
+            LOG.debug(
+                    "End of input subtask={} sourceIndex={} {}",
+                    readerContext.getIndexOfSubtask(),
+                    currentSourceIndex,
+                    currentReader);
+            if (currentSourceIndex + 1 < realReaders.size()) {
+                // Signal the coordinator that the current reader has consumed all input and the
+                // next source can potentially be activated (after all readers are ready).
+                readerContext.sendSourceEventToCoordinator(
+                        new SourceReaderFinishedEvent(currentSourceIndex, lastCheckpointId));
+                // More data will be available from the next reader.
+                // InputStatus.NOTHING_AVAILABLE requires us to complete the availability
+                // future after source switch to resume poll.
+                return InputStatus.NOTHING_AVAILABLE;
+            }
+        }
+        return status;
+    }
+
+    @Override
+    public List<HybridSourceSplit<?>> snapshotState(long checkpointId) {
+        this.lastCheckpointId = checkpointId;
+        List<? extends SourceSplit> state = currentReader.snapshotState(checkpointId);
+        return wrappedSplits(currentSourceIndex, state);
+    }
+
+    public static List<HybridSourceSplit<?>> wrappedSplits(
+            int readerIndex, List<? extends SourceSplit> state) {
+        List<HybridSourceSplit<?>> wrappedSplits = new ArrayList<>(state.size());
+        for (SourceSplit split : state) {
+            wrappedSplits.add(new HybridSourceSplit<>(readerIndex, split));
+        }
+        return wrappedSplits;
+    }
+
+    public static <SplitT extends SourceSplit> List<SplitT> unwrappedSplits(
+            List<HybridSourceSplit<SplitT>> splits) {
+        List<SplitT> unwrappedSplits = new ArrayList<>(splits.size());
+        for (HybridSourceSplit<SplitT> split : splits) {
+            unwrappedSplits.add(split.getWrappedSplit());
+        }
+        return unwrappedSplits;
+    }
+
+    @Override
+    public CompletableFuture<Void> isAvailable() {
+        return availabilityFuture = currentReader.isAvailable();
```

Comment:
Why not directly return `currentReader.isAvailable()`?

Reply:
added comment